Swap CeleryExecutor over to use TaskSDK for execution.

ashb opened the pull request:
Closes #45426

Some points of note about this PR:

- Logging is changed in Celery, but only for Airflow 3

  Celery does it's own "capture stdout" logging, which conflicts with the ones
  we do in the TaskSDK, so we disable that; but to not change anything for
  Airflow 3.

- Simplify task SDK logging redirection

  As part of this discovery that Celery captures stdout/stderr itself (and
  before disabling that) I discovered a simpler way to re-open the
  stdin/out/err so that the implementation needs fewer/no special casing.

- Make JSON task logs more readable by giving them a consistent/useful order

  We re-order (by re-creating) the event_dict so that timestamp, level, and
  then even are always the first items in the dict

- Makes the CeleryExecutor understand the concept of "workloads" instead a
  command tuple.

  This change isn't done in the best way, but until Kube executor is swapped
  over (and possibly the other in-tree executors, such as ECS) we need to
  support both styles concurrently.

  The change should be done in such a way that the provider still works with
  Airflow v2, if it's running on that version.

- Upgrade Celery

  This turned out to not be 100% necessary but it does fix some deprecation
  warnings when running on Python 3.12

- Ensure that the forked process in TaskSDK _never ever_ exits

  Again, this isn't possible usually, but since the setup step of `_fork_main`
  died, it didn't call `os._exit()`, and was caught further up, which meant
  the process stayed alive as it never closed the sockets properly. We put and
  extra safety try/except block in place to catch that

I have not yet included a newsfragment for changing the executor interface as
the old style is _currently_ still supported.


Testing: I ran `airflow scheduler`, `airflow fastapi-api`, and `airflow celery
worker` and triggered some dags, viewed logs etc.

And since a picture makes all PRs better

<img width="1550" alt="Screenshot 2025-01-29 at 22 49 49" src="https://github.com/user-attachments/assets/a6b1bc3c-917c-408e-bc7b-e4ac2890804e" />  


Celery window showing debug logs from the supervisor itself (task output goes to file)

<img width="1022" alt="Screenshot 2025-01-29 at 22 50 56" src="https://github.com/user-attachments/assets/922b801b-2f4a-4a69-a02d-dd9b536b7bd5" />

ashb commented on providers/celery/src/airflow/providers/celery/executors/celery_executor_utils.py:
```diff
@@ -125,21 +126,54 @@ def on_celery_import_modules(*args, **kwargs):
-    dag_id, task_id = BaseExecutor.validate_airflow_tasks_run_command(command_to_exec)
+# Once Celery5 is out of beta, we can pass `pydantic=True` to the decorator and it will handle the validation
+# and deserialization for us
+@app.task(name="execute_workload")
```
This is a small thing, but by specifying the name here, it means this is what shows up in the Celery logs, rather than `airflow.providers.celery.executors.celery_executor_utils.execute_workload`!

amoghrajesh replied:
Cool, damn nice!

ashb commented on providers/celery/src/airflow/providers/celery/executors/celery_executor_utils.py:
```diff
@@ -125,21 +126,54 @@ def on_celery_import_modules(*args, **kwargs):
+    )
+
+
+if not AIRFLOW_V_3_0_PLUS:
+
+    @app.task
+    def execute_command(command_to_exec: CommandType) -> None:
```
Note: as mentioned in the PR description, when running on Airflow 3, we don't register the old task anymore

ashb commented on providers/celery/src/airflow/providers/celery/executors/celery_executor_utils.py:
```diff
@@ -213,15 +247,15 @@ def send_task_to_executor(
     try:
         with timeout(seconds=OPERATION_TIMEOUT):
-            result = task_to_run.apply_async(args=[command], queue=queue)
+            result = task_to_run.apply_async(args=args, queue=queue)
```
This change lets this function be used by v2 and v3

ashb commented on providers/celery/src/airflow/providers/celery/executors/default_celery.py:
```diff
@@ -67,7 +68,7 @@ def _broker_supports_visibility_timeout(url):
 else:
     log.debug("Value for celery result_backend not found. Using sql_alchemy_conn with db+ prefix.")
-    result_backend = f'db+{conf.get("database", "SQL_ALCHEMY_CONN")}'
+    result_backend = f"db+{conf.get('database', 'SQL_ALCHEMY_CONN')}"
```
Ruff made this change for me (🤷🏻 since I'm touching this file anyway)

ashb commented on providers/celery/src/airflow/providers/celery/executors/default_celery.py:
```diff
@@ -81,6 +82,9 @@ def _broker_supports_visibility_timeout(url):
     "broker_url": broker_url,
     "broker_transport_options": broker_transport_options,
+    "broker_connection_retry_on_startup": conf.getboolean(
+        "celery", "broker_connection_retry_on_startup", fallback=True
```
This silences a warning on Celery 5.4

ashb commented on providers/celery/src/airflow/providers/celery/executors/default_celery.py:
```diff
@@ -126,9 +135,7 @@ def _get_celery_ssl_active() -> bool:
-        "AirflowConfigException: SSL_ACTIVE is True, "
-        "please ensure SSL_KEY, "
-        "SSL_CERT and SSL_CACERT are set"
+        "AirflowConfigException: SSL_ACTIVE is True, please ensure SSL_KEY, SSL_CERT and SSL_CACERT are set"
```
Ditto, ruff did it

amoghrajesh replied:
Haha!

ashb commented on task_sdk/src/airflow/sdk/log.py:
```diff
@@ -197,9 +191,19 @@ def logging_processors(
             exc_group_processor = None
 
         def json_dumps(msg, default):
+            # Note: this is _very_ expensive, but lets massage the dict order for nice viewing of the raw JSON
```
```suggestion
            # Note: this is likely an "expensive" step, but lets massage the dict order for nice
            # viewing of the raw JSON logs.
```

ashb commented on providers/celery/tests/provider_tests/celery/executors/test_celery_executor.py:
```diff
@@ -147,43 +147,6 @@ def test_gauge_executor_metrics(self, mock_stats_gauge, mock_trigger_tasks, mock
-            pytest.param(["airflow", "tasks", "run"], False, id="complete-command"),
-        ],
-    )
-    def test_command_validation(self, command, raise_exception):
```
This test is gone as we don't need to worry about this anymore

jedcunningham replied:
Won't we want it for the old core provider compat tests? Maybe should be kept but put behind a skipif(AF3...)?

amoghrajesh replied:
+1 on what Jed has to say, we will still continue to run the compat tests

jedcunningham commented on providers/celery/src/airflow/providers/celery/executors/celery_executor_utils.py:
```diff
@@ -125,21 +130,54 @@ def on_celery_import_modules(*args, **kwargs):
-def execute_command(command_to_exec: CommandType) -> None:
-    """Execute command."""
-    dag_id, task_id = BaseExecutor.validate_airflow_tasks_run_command(command_to_exec)
+# Once Celery5 is out of beta, we can pass `pydantic=True` to the decorator and it will handle the validation
```
Do you mean celery 5.5?

amoghrajesh replied:
Lets add a TODO?

amoghrajesh commented on airflow/executors/base_executor.py:
```diff
@@ -332,12 +351,12 @@ def trigger_tasks(self, open_slots: int) -> None:
 
         :param open_slots: Number of open slots
         """
-        span = Trace.get_current_span()
```
Do we not want to have the tracer anymore?

o-nikolas replied:
l'm also curious about the state of this one.

amoghrajesh commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -849,6 +848,9 @@ def process_executor_events(
                     continue
                 ti.task = task
                 if task.on_retry_callback or task.on_failure_callback:
+                    # Only log the error/extra info here, since the `ti.handle_failure()` path will log it
+                    # too, which would lead to double logging
+                    cls.logger().error(msg)
```
Nice!

amoghrajesh commented on providers/celery/src/airflow/providers/celery/executors/celery_executor.py:
```diff
@@ -228,6 +231,11 @@ class CeleryExecutor(BaseExecutor):
 
+    if TYPE_CHECKING and AIRFLOW_V_3_0_PLUS:
+        # In the v3 path, we store workloads, not commands as strings.
+        # TODO: TaskSDK: move this type change into BaseExecutor
```
```suggestion
        # TODO: TaskSDK: move this type change into BaseExecutor once all executors are ported
```

o-nikolas replied:
+1

amoghrajesh commented on providers/celery/src/airflow/providers/celery/executors/celery_executor.py:
```diff
@@ -256,10 +264,25 @@ def _num_tasks_per_send_process(self, to_send_count: int) -> int:
+        ]
+        self._send_tasks(tasks)
+
+    def _send_tasks(self, task_tuples_to_send):
```
Can we add typing?
Its probably: `list[TaskTuple])`

amoghrajesh commented on providers/celery/src/airflow/providers/celery/executors/celery_executor.py:
```diff
@@ -256,10 +264,25 @@ def _num_tasks_per_send_process(self, to_send_count: int) -> int:
+        self._send_tasks(tasks)
+
+    def _send_tasks(self, task_tuples_to_send):
+        first_task = next(t[-1] for t in task_tuples_to_send)
```
Simple change but much cleaner!

amoghrajesh commented on providers/celery/src/airflow/providers/celery/executors/celery_executor.py:
```diff
@@ -416,6 +439,10 @@ def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[Task
+            if isinstance(result.result, BaseException):
+                e = result.result
+                # Log the exception we got from the remote end
+                self.log.warning("Task %s tailed with error", ti.key, exc_info=e)
```
```suggestion
                self.log.warning("Task %s failed with error", ti.key, exc_info=e)
```

amoghrajesh commented on providers/celery/src/airflow/providers/celery/executors/celery_executor_utils.py:
```diff
@@ -125,21 +130,54 @@ def on_celery_import_modules(*args, **kwargs):
-            Stats.incr("celery.execute_command.failure")
-            raise
+
+    if not isinstance(workload, workloads.ExecuteTask):
+        raise ValueError(f"CeleryExecutor does not now how to handle {type(workload)}")
```
We can do this before trying to decode right?

ashb replied:
No, since before decoding we have a json string

amoghrajesh commented on providers/celery/src/airflow/providers/celery/executors/celery_executor_utils.py:
```diff
@@ -125,21 +130,54 @@ def on_celery_import_modules(*args, **kwargs):
+    from airflow.executors import workloads
+    from airflow.sdk.execution_time.supervisor import supervise
+
+    decoder = TypeAdapter(workloads.All)
+    workload = decoder.validate_json(input)
```
Should this rather be in a try block?
```
            try:
                workload = decoder.validate_json(input)
```

ashb replied:
I don't think so - we want it to fail so I don't know what we'd put in the except block other than raise

amoghrajesh replied:
Yeah doesnt add any value

amoghrajesh commented on task_sdk/src/airflow/sdk/execution_time/supervisor.py:
```diff
@@ -352,7 +341,17 @@ def start(
+                _fork_main(child_stdin, child_stdout, child_stderr, child_logs.fileno(), target)
+            except BaseException as e:
+                try:
+                    print("Exception in _fork_main, exiting with code 124", e, file=sys.stderr)
```
Lets log instead?

ashb replied:
We can't log, not reliably as this is such an unusual occurrence that something went wrong we can't rely on log being usable here

ashb replied:
I've added a comment about this

amoghrajesh commented on task_sdk/src/airflow/sdk/log.py:
```diff
@@ -197,9 +193,19 @@ def logging_processors(
+            # Note: this is likely an "expensive" step, but lets massage the dict order for nice
+            # viewing of the raw JSON logs.
+            # Maybe we don't need this once the UI renders the JSON instead of displaying the raw text
+            msg = {
+                "timestamp": msg.pop("timestamp"),
+                "level": msg.pop("level"),
+                "event": msg.pop("event"),
+                **msg,
+            }
```
This is super useful, I have been meaning to get to it, as it was just harder to see the "event" normally. +1 for this change

amoghrajesh reviewed (commented):
Thanks for working on this @ashb!
Have a few comments, nothing serious, general comments / nits, I am OK with the PR mostly

ashb committed:
Swap CeleryExecutor over to use TaskSDK for execution.

Some points of note about this change:

- Logging is changed in Celery, but only for Airflow 3

  Celery does it's own "capture stdout" logging, which conflicts with the ones
  we do in the TaskSDK, so we disable that; but to not change anything for
  Airflow 3.

- Simplify task SDK logging redirection

  As part of this discovery that Celery captures stdout/stderr itself (and
  before disabling that) I discovered a simpler way to re-open the
  stdin/out/err so that the implementation needs fewer/no special casing.

- Make JSON task logs more readable by giving them a consistent/useful order

  We re-order (by re-creating) the event_dict so that timestamp, level, and
  then even are always the first items in the dict

- Makes the CeleryExecutor understand the concept of "workloads" instead a
  command tuple.

  This change isn't done in the best way, but until Kube executor is swapped
  over (and possibly the other in-tree executors, such as ECS) we need to
  support both styles concurrently.

  The change should be done in such a way that the provider still works with
  Airflow v2, if it's running on that version.

- Upgrade Celery

  This turned out to not be 100% necessary but it does fix some deprecation
  warnings when running on Python 3.12

- Ensure that the forked process in TaskSDK _never ever_ exits

  Again, this isn't possible usually, but since the setup step of `_fork_main`
  died, it didn't call `os._exit()`, and was caught further up, which meant
  the process stayed alive as it never closed the sockets properly. We put and
  extra safety try/except block in place to catch that

I have not yet included a newsfragment for changing the executor interface as
the old style is _currently_ still supported.

o-nikolas commented on tests/integration/executors/test_celery_executor.py:
```diff
@@ -181,54 +178,47 @@ def fake_execute_command(command):
-        with _prepare_app(execute=fake_execute_command):
-            # fake_execute_command takes no arguments while execute_command takes 1,
+        with _prepare_app(execute=fake_task):
+            # fake_execute_command takes no arguments while execute_workload takes 1,
```
```suggestion
            # fake_task takes no arguments while execute_workload takes 1,
```

o-nikolas reviewed (approved):
Changes to base look good to me to support both for now. I think it's called out in a comment but we should migrate all in-tree executors (K8s, ECS, Batch, Edge, etc) and add a new min Airflow version for those packages.

jscheffl reviewed (approved):
Wow I would have expected a larger PR from the challenges you faced. Looks pretty solid.

Did a regression with EdgeExecutor and also this is still working. Just one detail - when I executed my "Integration Test" DAG from the Edge Examples, the BranchOperator in EdgeWorker is failing, which seems to be working in Celery. So might be I need a bit of ENV fixing. Exception is:
{"timestamp":"2025-01-30T22:01:35.193579","level":"error","event":"Task failed with exception","logger":"task","error_detail":[{"exc_type":"OperationalError","exc_value":"(sqlite3.OperationalError) no such table: dag_run\n[SQL: SELECT dag_run.state AS dag_run_state, dag_run.id AS dag_run_id, dag_run.dag_id AS dag_run_dag_id, dag_run.queued_at AS dag_run_queued_at, dag_run.logical_date AS dag_run_logical_date, dag_run.start_date AS dag_run_start_date, dag_run.end_date AS dag_run_end_date, dag_run.run_id AS dag_run_run_id, dag_run.creating_job_id AS dag_run_creating_job_id, dag_run.external_trigger AS dag_run_external_trigger, dag_run.run_type AS dag_run_run_type, dag_run.triggered_by AS dag_run_triggered_by, dag_run.conf AS dag_run_conf, dag_run.data_interval_start AS dag_run_data_interval_start, dag_run.data_interval_end AS dag_run_data_interval_end, dag_run.last_scheduling_decision AS dag_run_last_scheduling_decision, dag_run.log_template_id AS dag_run_log_template_id, dag_run.updated_at AS dag_run_updated_at, dag_run.clear_number AS dag_run_clear_number, dag_run.backfill_id AS dag_run_backfill_id, dag_run.dag_version_id AS dag_run_dag_version_id, dag_run.bundle_version AS dag_run_bundle_version \nFROM dag_run \nWHERE dag_run.dag_id = ? AND dag_run.run_id = ?]\n[parameters: ('integration_test', 'manual__2025-01-30T23:01:24+01:00')]\n(Background on this error at: [https://sqlalche.me/e/14/e3q8)","syntax_error":null,"is_cause":false,"frames":[](https://sqlalche.me/e/14/e3q8)%22,%22syntax_error%22:null,%22is_cause%22:false,%22frames%22:[){"filename":"/opt/airflow/task_sdk/src/airflow/sdk/execution_time/task_runner.py","lineno":527,"name":"run"},{"filename":"/opt/airflow/task_sdk/src/airflow/sdk/execution_time/task_runner.py","lineno":626,"name":"_execute_task"},{"filename":"/opt/airflow/airflow/models/baseoperator.py","lineno":173,"name":"wrapper"},{"filename":"/opt/airflow/airflow/decorators/base.py","lineno":252,"name":"execute"},{"filename":"/opt/airflow/airflow/models/baseoperator.py","lineno":173,"name":"wrapper"},{"filename":"/opt/airflow/providers/standard/src/airflow/providers/standard/operators/python.py","lineno":240,"name":"execute"},{"filename":"/opt/airflow/airflow/operators/branch.py","lineno":41,"name":"do_branch"},{"filename":"/opt/airflow/airflow/models/skipmixin.py","lineno":212,"name":"skip_all_except"},{"filename":"/opt/airflow/airflow/utils/session.py","lineno":98,"name":"wrapper"},{"filename":"/opt/airflow/airflow/models/taskinstance.py","lineno":2480,"name":"get_dagrun"},{"filename":"/opt/airflow/airflow/models/taskinstance.py","lineno":2461,"name":"_get_dagrun"},{"filename":"/usr/local/lib/python3.12/site-packages/sqlalchemy/orm/query.py","lineno":2870,"name":"one"},{"filename":"/usr/local/lib/python3.12/site-packages/sqlalchemy/orm/query.py","lineno":2916,"name":"_iter"},{"filename":"/usr/local/lib/python3.12/site-packages/sqlalchemy/orm/session.py","lineno":1717,"name":"execute"},{"filename":"/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py","lineno":1710,"name":"_execute_20"},{"filename":"/usr/local/lib/python3.12/site-packages/sqlalchemy/sql/elements.py","lineno":334,"name":"_execute_on_connection"},{"filename":"/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py","lineno":1577,"name":"_execute_clauseelement"},{"filename":"/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py","lineno":1953,"name":"_execute_context"},{"filename":"/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py","lineno":2134,"name":"_handle_dbapi_exception"},{"filename":"/usr/local/lib/python3.12/site-packages/sqlalchemy/util/compat.py","lineno":211,"name":"raise_"},{"filename":"/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py","lineno":1910,"name":"_execute_context"},{"filename":"/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/default.py","lineno":736,"name":"do_execute"}]},{"exc_type":"OperationalError","exc_value":"no such table: dag_run","syntax_error":null,"is_cause":true,"frames":[{"filename":"/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py","lineno":1910,"name":"_execute_context"},{"filename":"/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/default.py","lineno":736,"name":"do_execute"}]}]}

..but anyway, LGTM!

amoghrajesh reviewed (approved):
Thanks for addressing the comments, LGTM +1
